File: src/lib/engine/ruleEnforcer.ts
import type { Location, StateChange, TurnResult, WorldState } from "./types";

/**
 * 寬鬆 normalize：去除所有空白、常見標點、轉小寫。
 * 讓「黃藍紅」對上「黃 藍 紅」、「B,A,C」對上「B A C」。
 */
function normalizeSolution(s: string): string {
  return s
    .trim()
    .toLowerCase()
    .replace(/[\s、,，。.\-_]/g, ""); // 去空白 + 中英標點 + 底線
}

/**
 * Return a list of violation messages. Empty list means all changes are valid.
 */
export function validate(worldState: WorldState, turnResult: TurnResult): string[] {
  const violations: string[] = [];
  const locIds = new Set(Object.keys(worldState.locations));
  const itemIds = new Set(Object.keys(worldState.items));
  const currentLoc = worldState.locations[worldState.currentLocationId];

  for (const sc of turnResult.stateChanges) {
    violations.push(...validateChange(sc, worldState, currentLoc, locIds, itemIds));
  }

  return violations;
}

function validateChange(
  sc: StateChange,
  ws: WorldState,
  currentLoc: Location,
  locIds: Set<string>,
  itemIds: Set<string>,
): string[] {
  const v: string[] = [];

  if (sc.type === "move_player") {
    if (!sc.toLocation) {
      v.push("move_player missing to_location");
    } else if (!locIds.has(sc.toLocation)) {
      v.push(`move_player: unknown location '${sc.toLocation}'`);
    } else if (!currentLoc.connectedLocationIds.includes(sc.toLocation)) {
      v.push(
        `move_player: '${sc.toLocation}' is not connected to '${ws.currentLocationId}'`,
      );
    } else {
      const destLoc = ws.locations[sc.toLocation];
      const unsolvedLocks = (destLoc?.lockedByPuzzleIds ?? []).filter(
        (pid) => !ws.puzzles[pid]?.isSolved,
      );
      if (unsolvedLocks.length > 0) {
        v.push(
          `move_player: '${sc.toLocation}' is locked — solve puzzles first: ${unsolvedLocks.join(", ")}`,
        );
      }
    }
  } else if (sc.type === "take_item") {
    if (!sc.itemId) {
      v.push("take_item missing item_id");
    } else if (!itemIds.has(sc.itemId)) {
      v.push(`take_item: unknown item '${sc.itemId}'`);
    } else {
      const item = ws.items[sc.itemId];
      if (item.locationId !== ws.currentLocationId) {
        v.push(`take_item: '${sc.itemId}' is not in current location (item.locationId='${item.locationId}')`);
      } else if (ws.inventory.includes(sc.itemId)) {
        v.push(`take_item: '${sc.itemId}' already in inventory`);
      } else if (!item.isTakeable) {
        v.push(`take_item: '${sc.itemId}' is not takeable`);
      } else if (item.isLocked) {
        if (!item.unlockItemId || !ws.inventory.includes(item.unlockItemId)) {
          v.push(
            `take_item: '${sc.itemId}' is locked — need '${item.unlockItemId}' in inventory`,
          );
        }
      }
    }
  } else if (sc.type === "use_item") {
    if (!sc.itemId) {
      v.push("use_item missing item_id");
    } else if (!itemIds.has(sc.itemId)) {
      v.push(`use_item: unknown item '${sc.itemId}'`);
    } else if (!ws.inventory.includes(sc.itemId)) {
      v.push(`use_item: '${sc.itemId}' not in inventory`);
    }
  } else if (sc.type === "move_item") {
    if (!sc.itemId) {
      v.push("move_item missing item_id");
    } else if (!itemIds.has(sc.itemId)) {
      v.push(`move_item: unknown item '${sc.itemId}'`);
    } else if (!currentLoc.itemIds.includes(sc.itemId)) {
      v.push(`move_item: '${sc.itemId}' is not in current location`);
    } else if (!sc.toLocation || !locIds.has(sc.toLocation)) {
      v.push(`move_item: invalid to_location '${sc.toLocation}'`);
    }
  } else if (sc.type === "solve_puzzle") {
    if (!sc.puzzleId) {
      v.push("solve_puzzle missing puzzle_id");
    } else if (!(sc.puzzleId in ws.puzzles)) {
      v.push(`solve_puzzle: unknown puzzle '${sc.puzzleId}'`);
    } else {
      const puzzle = ws.puzzles[sc.puzzleId];
      if (puzzle.locationId !== ws.currentLocationId) {
        v.push(`solve_puzzle: puzzle '${sc.puzzleId}' is not in current location`);
      } else if (puzzle.isSolved) {
        v.push(`solve_puzzle: puzzle '${sc.puzzleId}' already solved`);
      } else if (
        !sc.attemptedSolution ||
        normalizeSolution(sc.attemptedSolution) !== normalizeSolution(puzzle.solution)
      ) {
        console.warn(
          `[solve_puzzle wrong-solution] puzzle=${sc.puzzleId}` +
            ` attempted_raw=${JSON.stringify(sc.attemptedSolution)}` +
            ` attempted_norm=${sc.attemptedSolution ? normalizeSolution(sc.attemptedSolution) : "null"}` +
            ` solution_norm=${normalizeSolution(puzzle.solution)}`,
        );
        v.push(`solve_puzzle: wrong solution for puzzle '${sc.puzzleId}'`);
      }
    }
  }

  return v;
}

/**
 * Apply all state changes and return a new WorldState. Assumes validate() returned [].
 * action: the player's raw input for this turn — stored in history so the LLM
 * can see Player: lines in subsequent turns.
 */
export function apply(worldState: WorldState, turnResult: TurnResult, action = ""): WorldState {
  const ws = structuredClone(worldState);

  for (const sc of turnResult.stateChanges) {
    applyChange(ws, sc);
  }

  ws.isWon = checkWin(ws);
  ws.turnCount += 1;
  appendHistory(ws, turnResult.narration, action);
  return ws;
}

function applyChange(ws: WorldState, sc: StateChange): void {
  if (sc.type === "move_player") {
    ws.currentLocationId = sc.toLocation!;
  } else if (sc.type === "take_item") {
    const iid = sc.itemId!;
    const loc = ws.locations[ws.currentLocationId];
    const idx = loc.itemIds.indexOf(iid);
    if (idx !== -1) loc.itemIds.splice(idx, 1);
    ws.items[iid].locationId = "inventory";
    if (!ws.inventory.includes(iid)) ws.inventory.push(iid);
  } else if (sc.type === "use_item") {
    // narration-only side effect; enforcer already verified it's in inventory
  } else if (sc.type === "move_item") {
    const iid = sc.itemId!;
    const fromLoc = ws.locations[ws.currentLocationId];
    const idx = fromLoc.itemIds.indexOf(iid);
    if (idx !== -1) fromLoc.itemIds.splice(idx, 1);
    const dest = ws.locations[sc.toLocation!];
    if (!dest.itemIds.includes(iid)) dest.itemIds.push(iid);
    ws.items[iid].locationId = sc.toLocation!;
  } else if (sc.type === "solve_puzzle") {
    const puzzle = ws.puzzles[sc.puzzleId!];
    puzzle.isSolved = true;
    if (puzzle.rewardItemId) {
      const loc = ws.locations[ws.currentLocationId];
      if (!loc.itemIds.includes(puzzle.rewardItemId)) loc.itemIds.push(puzzle.rewardItemId);
      ws.items[puzzle.rewardItemId].locationId = ws.currentLocationId;
    }
  }
}

function checkWin(ws: WorldState): boolean {
  if (ws.currentLocationId !== ws.winCondition.targetLocationId) return false;
  return ws.winCondition.requiredSolvedPuzzleIds.every((pid) => ws.puzzles[pid].isSolved);
}

function appendHistory(ws: WorldState, narration: string, action = ""): void {
  // No cap here — full history is kept in DB so players can pick any narration as a
  // share-card quote.  LLM context is capped separately in turnHandler (MAX_HISTORY=15).
  ws.history.push({ action, narration });
}

/** Return a new WorldState with 'Nothing happens.' recorded and turn incremented. */
export function applyFallback(worldState: WorldState, action: string): WorldState {
  const ws = structuredClone(worldState);
  ws.turnCount += 1;
  appendHistory(ws, "Nothing happens.", action);
  return ws;
}

/**
 * Validate a single StateChange and, if valid, return the state with it applied.
 * Does not increment turnCount or append history — callers use apply() for full turn semantics.
 */
export function enforceStateChange(
  state: WorldState,
  change: StateChange,
): { valid: boolean; reason?: string; updatedState?: WorldState } {
  const locIds = new Set(Object.keys(state.locations));
  const itemIds = new Set(Object.keys(state.items));
  const currentLoc = state.locations[state.currentLocationId];

  const violations = validateChange(change, state, currentLoc, locIds, itemIds);
  if (violations.length > 0) {
    return { valid: false, reason: violations[0] };
  }

  const ws = structuredClone(state);
  applyChange(ws, change);
  return { valid: true, updatedState: ws };
}

File: reference/engine/rule_enforcer.py
"""Pure-Python validator and state applier. No I/O, no async."""
from __future__ import annotations

import copy

from .models import StateChange, TurnResult, WorldState

MAX_HISTORY = 15


def validate(world_state: WorldState, turn_result: TurnResult) -> list[str]:
    """Return a list of violation messages. Empty list means all changes are valid."""
    violations: list[str] = []
    loc_ids = set(world_state.locations)
    item_ids = set(world_state.items)
    current_loc = world_state.locations[world_state.current_location_id]

    for sc in turn_result.state_changes:
        violations.extend(_validate_change(sc, world_state, current_loc, loc_ids, item_ids))

    return violations


def _validate_change(
    sc: StateChange,
    ws: WorldState,
    current_loc,
    loc_ids: set[str],
    item_ids: set[str],
) -> list[str]:
    v: list[str] = []

    if sc.type == "move_player":
        if not sc.to_location:
            v.append("move_player missing to_location")
        elif sc.to_location not in loc_ids:
            v.append(f"move_player: unknown location '{sc.to_location}'")
        elif sc.to_location not in current_loc.connected_location_ids:
            v.append(
                f"move_player: '{sc.to_location}' is not connected to '{ws.current_location_id}'"
            )

    elif sc.type == "take_item":
        if not sc.item_id:
            v.append("take_item missing item_id")
        elif sc.item_id not in item_ids:
            v.append(f"take_item: unknown item '{sc.item_id}'")
        else:
            item = ws.items[sc.item_id]
            if sc.item_id not in current_loc.item_ids:
                v.append(f"take_item: '{sc.item_id}' is not in current location")
            elif sc.item_id in ws.inventory:
                v.append(f"take_item: '{sc.item_id}' already in inventory")
            elif not item.is_takeable:
                v.append(f"take_item: '{sc.item_id}' is not takeable")
            elif item.is_locked:
                if not item.unlock_item_id or item.unlock_item_id not in ws.inventory:
                    v.append(
                        f"take_item: '{sc.item_id}' is locked — need '{item.unlock_item_id}' in inventory"
                    )

    elif sc.type == "use_item":
        if not sc.item_id:
            v.append("use_item missing item_id")
        elif sc.item_id not in item_ids:
            v.append(f"use_item: unknown item '{sc.item_id}'")
        elif sc.item_id not in ws.inventory:
            v.append(f"use_item: '{sc.item_id}' not in inventory")

    elif sc.type == "move_item":
        if not sc.item_id:
            v.append("move_item missing item_id")
        elif sc.item_id not in item_ids:
            v.append(f"move_item: unknown item '{sc.item_id}'")
        elif sc.item_id not in current_loc.item_ids:
            v.append(f"move_item: '{sc.item_id}' is not in current location")
        elif not sc.to_location or sc.to_location not in loc_ids:
            v.append(f"move_item: invalid to_location '{sc.to_location}'")

    elif sc.type == "solve_puzzle":
        if not sc.puzzle_id:
            v.append("solve_puzzle missing puzzle_id")
        elif sc.puzzle_id not in ws.puzzles:
            v.append(f"solve_puzzle: unknown puzzle '{sc.puzzle_id}'")
        else:
            puzzle = ws.puzzles[sc.puzzle_id]
            if puzzle.location_id != ws.current_location_id:
                v.append(
                    f"solve_puzzle: puzzle '{sc.puzzle_id}' is not in current location"
                )
            elif puzzle.is_solved:
                v.append(f"solve_puzzle: puzzle '{sc.puzzle_id}' already solved")
            elif not sc.attempted_solution or sc.attempted_solution.strip().lower() != puzzle.solution.strip().lower():
                v.append(
                    f"solve_puzzle: wrong solution for puzzle '{sc.puzzle_id}'"
                )

    return v


def apply(world_state: WorldState, turn_result: TurnResult) -> WorldState:
    """Apply all state changes and return a new WorldState. Assumes validate() returned []."""
    ws = world_state.model_copy(deep=True)

    for sc in turn_result.state_changes:
        _apply_change(ws, sc)

    ws.is_won = _check_win(ws)
    ws.turn_count += 1
    _append_history(ws, turn_result.narration)
    return ws


def _apply_change(ws: WorldState, sc: StateChange) -> None:
    if sc.type == "move_player":
        ws.current_location_id = sc.to_location  # type: ignore[assignment]

    elif sc.type == "take_item":
        iid = sc.item_id
        loc = ws.locations[ws.current_location_id]
        if iid in loc.item_ids:
            loc.item_ids.remove(iid)
        ws.items[iid].location_id = "inventory"  # type: ignore[index]
        if iid not in ws.inventory:
            ws.inventory.append(iid)  # type: ignore[arg-type]

    elif sc.type == "use_item":
        pass  # narration-only side effect; enforcer already verified it's in inventory

    elif sc.type == "move_item":
        iid = sc.item_id
        from_loc = ws.locations[ws.current_location_id]
        if iid in from_loc.item_ids:
            from_loc.item_ids.remove(iid)
        dest = ws.locations[sc.to_location]  # type: ignore[index]
        if iid not in dest.item_ids:
            dest.item_ids.append(iid)
        ws.items[iid].location_id = sc.to_location  # type: ignore[assignment,index]

    elif sc.type == "solve_puzzle":
        puzzle = ws.puzzles[sc.puzzle_id]  # type: ignore[index]
        puzzle.is_solved = True
        if puzzle.reward_item_id:
            loc = ws.locations[ws.current_location_id]
            if puzzle.reward_item_id not in loc.item_ids:
                loc.item_ids.append(puzzle.reward_item_id)
            ws.items[puzzle.reward_item_id].location_id = ws.current_location_id


def _check_win(ws: WorldState) -> bool:
    if ws.current_location_id != ws.win_condition.target_location_id:
        return False
    return all(
        ws.puzzles[pid].is_solved
        for pid in ws.win_condition.required_solved_puzzle_ids
    )


def _append_history(ws: WorldState, narration: str, action: str = "") -> None:
    ws.history.append({"action": action, "narration": narration})
    if len(ws.history) > MAX_HISTORY:
        ws.history = ws.history[-MAX_HISTORY:]


def apply_fallback(world_state: WorldState, action: str) -> WorldState:
    """Return a new WorldState with 'Nothing happens.' recorded and turn incremented."""
    ws = world_state.model_copy(deep=True)
    ws.turn_count += 1
    _append_history(ws, "Nothing happens.", action)
    return ws
